### PR TITLE
patching the delmailuser script to function properly (+ refactoring)

### LINKED ISSUE #1808: [BUG] Mailbox contents not deleted
# Bug Report

## Context
Mailbox contents remain in `maildata` subdirectory after running `sudo ./setup.sh email del test@example.com`.

### What is affected by this bug?
Deleting mailboxes.
### When does this occur?
When deleting a mailbox with the `setup.sh` script.
### How do we replicate the issue?

```
sudo ./setup.sh email del test@example.com
Password: 
sed: can't read /tmp/docker-mailserver/postfix-virtual.cf: No such file or directory
Do you want to delete the mailbox as well(all mails will be removed)?(y/n) y

Mailbox deleted.
```
Check contents of `maildata` subdirectory.

### Actual Behavior
The mailbox is no longer listed when running `sudo ./setup.sh email list`. But the `test` mailbox and its contents are still in the `maildata` subdirectory. 
### Expected Behavior
The mailbox contents should have been deleted.
## Your Environment
Synology DS918+ with latest DSM and Docker version 18.09.8, build bfed4f5.

### Environment Variables

```yml
---
version: "2"

# Define services
services:
  
  roundcubemail:
    image: roundcube/roundcubemail:latest
    container_name: roundcubemail
    restart: unless-stopped
    volumes:
      - /v/d/roundcube/www:/var/www/html
      - /v/d/roundcube/sqlite:/var/roundcube/db
    ports:
      - 127.0.0.1:8000:80
    environment:
      - ROUNDCUBEMAIL_DB_TYPE=sqlite
      - ROUNDCUBEMAIL_SKIN=elastic
      - ROUNDCUBEMAIL_DEFAULT_HOST=ssl://mail.example.com
      - ROUNDCUBEMAIL_SMTP_SERVER=ssl://mail.example.com
      - ROUNDCUBEMAIL_DEFAULT_PORT=993
      - ROUNDCUBEMAIL_SMTP_PORT=465
  
  solr:
    container_name: mailserver_solr
    image: lmmdock/dovecot-solr:latest
    volumes:
      - /v/d/solr-dovecot:/opt/solr/server/solr/dovecot
    restart: always
  
  mailserver:
    image: docker.io/mailserver/docker-mailserver:latest
    hostname: mail
    domainname: example.com
    container_name: mailserver
    ports:
      - "465:465"
      - "993:993"
    volumes:
      - /v/d/mailserver/maildata:/var/mail
      - /v/d/mailserver/mailstate:/var/mail-state
      - /v/d/mailserver/maillogs:/var/log/mail
      - /v/d/mailserver/config:/tmp/docker-mailserver/
      - /v/d/ssl/output:/tmp/ssl:ro
      - /v/d/mailserver/solr-dovecot/10-plugin.conf:/etc/dovecot/conf.d/10-plugin.conf:ro
    environment:
      - SSL_TYPE=manual
      - SSL_CERT_PATH=/tmp/ssl/fullchain.pem
      - SSL_KEY_PATH=/tmp/ssl/privkey.pem
      - POSTFIX_MESSAGE_SIZE_LIMIT=102400000
      - ENABLE_SPAMASSASSIN=0
      - ENABLE_CLAMAV=0
      - ENABLE_FAIL2BAN=0
      - ENABLE_POSTGREY=0
      - ENABLE_SASLAUTHD=0
      - ONE_DIR=1
      - DMS_DEBUG=0
    cap_add:
      - NET_ADMIN
      - SYS_PTRACE
    restart: always
```
## Workaround?
Contents of mailbox are deleted when running `sudo docker exec mailserver rm -R /var/mail/example.com/test/` afterwards.

### PR DESCRIPTION
**Pull request by aendeavor**
# Description

This PR refactors `delmailuser` to function properly (hopefully).

Fixes #1808 (issue)

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (non-breaking change that does improve existing functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README.md or ENVIRONMENT.md or the Wiki)
- [x] If necessary I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by aendeavor** [comment]
So 

``` BASH
echo "Mailbox deleted." || errex "Mailbox couldn't be deleted: ${?}"
```

is obviously hilarious. The tests however do not seem to like

``` BASH
if [[ -d "/var/mail/${MAILARR[1]}/${MAILARR[0]}" ]]
then
  ...
else
  errex "Mailbox directory '/var/mail/${MAILARR[1]}/${MAILARR[0]}' does not exist." # <- especially this
fi
```

which before would not be exiting with a non-zero exit code. I would just change this to

``` BASH
if [[ -d "/var/mail/${MAILARR[1]}/${MAILARR[0]}" ]]
then
  ...
else
  echo "Mailbox directory '/var/mail/${MAILARR[1]}/${MAILARR[0]}' did not exist."
fi
```

@wernerfred Your take?

File: target/bin/delmailuser [hunk 1]
```diff
@@ -1,92 +1,108 @@
...
-    if sed -i -e "/ ""${USER}""$/d" \
-        -e "s/,""${USER}""//g" \
-        -e "s/""${USER}"",//g" "${ALIAS_DATABASE}"
+    if [[ -f ${ALIAS_DATABASE} ]]
     then
-      echo "${USER} and potential aliases deleted." || errex "Aliases for ${USER} couldn't be deleted in ${ALIAS_DATABASE}. ${?}"
+      # delete all aliases where the user is the only recipient( " ${USER$}" )
+      # delete user only for all aliases that deliver to multiple recipients ( ",${USER}" "${USER,}" )
+      if sed -i \
+        -e "/ ${USER}$/d" -e "s/,${USER}//g" -e "s/${USER},//g" \
```
Comment (wernerfred) [inline target/bin/delmailuser:72]: This is much better now 👍🏻

File: target/bin/delmailuser [hunk 2]
```diff
@@ -1,92 +1,108 @@
...
-      read -r -p "Do you want to delete the mailbox as well(all mails will be removed)?(y/n) " MAILDEL
-      echo
+      echo "Leaving the mailbox untouched. If you want to delete it at a later point use 'sudo docker exec mail rm -R /var/mail/${MAILARR[1]}/${MAILARR[0]}'"
+      exit 0
     fi
 
-    [[ ${MAILDEL} != "y" ]] && errex "Leaving the mailbox untouched. If you want to delete it at a later point use \"sudo docker exec mail rm -R /var/mail/${MAILARR[1]}/${MAILARR[0]}\""
-
-    if rm -r -f "/var/mail/${MAILARR[1]}/${MAILARR[0]}"
+    if [[ -d "/var/mail/${MAILARR[1]}/${MAILARR[0]}" ]]
```
Comment (wernerfred) [inline target/bin/delmailuser:96]: Will hardcoding `/var/mail/` work with setups that have a custom/different maildir set? Thinking about the PR we merged a few days ago (#1792)

**Review by wernerfred — APPROVED** [review]

**Comment by wernerfred** [comment]
> @wernerfred Your take? Ima be honest here, not sure if the tests are written to conform to the shitty script...?

Imho (no offense to the original test writers) for me it was really hard to understand what's going on with a lot of the tests I examined as I had to patch a test for the dkim PR. There are a lot of tests where calling them bad style would be too nice. Sometimes the authors even stated in the comments above that the test is really crap and will propably also pass if something else is given as ENV (testing only `wc -l` ). There would be a major refactoring necessary. Also there is a huge wild growth of configuration dirs and configs as every test author approached it a bit different. We should also consider writing a How-To about the tests as it is really hard to get into it from a outside perspective. 


> ```shell
> echo "Mailbox deleted." || errex "Mailbox couldn't be deleted: ${?}"
> ```

🤣 🤣 🤣

File: target/bin/delmailuser (hunk 2 above)
Comment (aendeavor) [reply target/bin/delmailuser:96]: This is most likely going to be a concern in the future - I think this PR shouldn't worry about this for now:)

**Comment by aendeavor** [comment]
> > @wernerfred Your take? Ima be honest here, not sure if the tests are written to conform to the shitty script...?
> 
> Speaking directly (no offense to the original test writers) for me it was really hard to understand what's going on with a lot of the tests I examined as I had to patch a test for the dkim PR. There are a lot of tests where calling them bad style would be too nice. Sometimes the authors even stated in the comments above that the test is really crap and will propably also pass if something else is given as ENV (testing only `wc -l` ). Imho there would be a major refactoring necessary. Also there is a huge wild growth of configuration dirs and configs as every test author approached it a bit different. We should also consider writing a How-To about the tests as it is really hard to get into it from a outside perspective.

Indeed, the tests will need complete overhaul in the future.
 
> > ```shell
> > echo "Mailbox deleted." || errex "Mailbox couldn't be deleted: ${?}"
> > ```
> 
> rofl rofl rofl

I didn't even see this the first time, but when I saw it, I was perplexed...

---

Tests are passing now. This is ready for merging, only waiting for your review.

**Review by wernerfred — APPROVED** [review]

**Pull request merged**